lib/stronger_parameters/controller_support/permitted_parameters.rb

```diff
@@ -0,0 +1,169 @@
+require 'stronger_parameters/constraints'
+
+module StrongerParameters
+  module ControllerSupport
+    module PermittedParameters
+      def self.included(klass)
+        klass.extend ClassMethods
+        klass.before_action :permit_parameters
+      end
+
+      class PermittedParametersHash < HashWithIndifferentAccess
+        def initialize(other = nil)
+          super()
+          merge!(other) unless other.nil?
+        end
+
+        def merge!(other)
+          return self if other.nil?
+
+          other.each do |key, value|
+            value = sugar(value)
+
+            if value.is_a?(StrongerParameters::Constraint)
+              self[key] = value
+            else
+              self[key] ||= self.class.new
+              self[key].merge!(value)
+            end
+          end
+
+          self
+        end
+
+        def merge(other)
+          dup.merge!(other)
+        end
+
+        def dup
+          super.tap do |duplicate|
+            duplicate.each do |k, v|
+              duplicate[k] = if v == true
+                true
+              else
+                v.dup
+              end
+            end
+          end
+        end
+
+        def sugar(value)
+          case value
+          when Array
+            ActionController::Parameters.array(*value.map { |v| sugar(v) }) | ActionController::Parameters.nil
```

Comment:
allowing nil for arrays seems sketchy ... we have a global `allow nil everywhere` option, so should not add another hardcoded one here ...